apps/src/lib/node/ledger/shell.rs

```diff
@@ -258,50 +262,87 @@ impl Shell {
             .expect("Must be able to update epoch");
     }
 
-    /// Validate and apply a transaction.
-    pub fn apply_tx(&mut self, req: request::DeliverTx) -> response::DeliverTx {
-        let mut response = response::DeliverTx::default();
-        let result = protocol::apply_tx(
-            &*req.tx,
-            &mut self.gas_meter,
-            &mut self.write_log,
-            &self.storage,
-        )
-        .map_err(Error::TxError);
+    // Verify the headers of the block
+    // TODO: This will include checking announcements of stake, individual and
+    // aggregated PVSS instances from the DKG protocol
+    pub fn verify_header(
+        &self,
+        _req: shim::request::VerifyHeader,
+    ) -> shim::response::VerifyHeader {
+        Default::default()
+    }
 
-        match result {
-            Ok(result) => {
-                if result.is_accepted() {
-                    tracing::info!(
-                        "all VPs accepted apply_tx storage modification {:#?}",
-                        result
-                    );
-                    self.write_log.commit_tx();
-                } else {
-                    tracing::info!(
-                        "some VPs rejected apply_tx storage modification {:#?}",
-                        result.vps_result.rejected_vps
-                    );
-                    self.write_log.drop_tx();
-                    response.code = 1;
-                }
-                response.gas_used = gas::as_i64(result.gas_used);
-                response.info = result.to_string();
-            }
-            Err(msg) => {
-                response.gas_used =
-                    gas::as_i64(self.gas_meter.get_current_transaction_gas());
-                response.info = msg.to_string();
-            }
-        }
-        response
+    /// Check the fees and signatures of the fee payer for a transaction
+    pub fn process_proposal(
+        &mut self,
+        _req: shim::request::ProcessProposal,
+    ) -> shim::response::ProcessProposal {
+        Default::default()
     }
 
-    /// End a block.
-    pub fn end_block(&mut self, _height: BlockHeight) -> response::EndBlock {
+    pub fn revert_proposal(
+        &mut self,
+        _req: shim::request::RevertProposal,
+    ) -> shim::response::RevertProposal {
         Default::default()
     }
 
+    pub fn extend_vote(
+        &mut self,
+        _req: shim::request::ExtendVote,
+    ) -> shim::response::ExtendVote {
+        Default::default()
+    }
+
+    /// Validate and apply transactions.
+    /// TODO: Will we also decrypt transactions here?
```

Comment:
I think we might decrypt to be decrypting them in `VerifyVoteExtension` and adding them them to a queue

Reply:
The queue is only necessary if we decide to encrypt gas limits right?

Reply:
yeah, I think we still have to have the queue